README.md
**Wastebin** is a web service that allows you to share notes anonymously, an alternative to `pastebin.com`.

This project was forked from [Lenpaste](https://git.lcomrade.su/root/lenpaste) and has been modified.

## Features

- No need to register
- Does not use cookies
- Can work without JavaScript
- Has its own API
- Open source and self-hosted

## Launch your own server

1. If you don't already have Docker installed, do so:

```bash
apt-get install -y docker docker.io docker-compose
```

2. Use a file like this `docker-compose.yml`:

```yaml
version: '3'

services:
  lenpaste:
    image: git.lcomrade.su/root/lenpaste:latest
    restart: always
    environment:
      # All parameters are optional
      - WASTEBIN_ADDRESS=:80                 # Set -address flag
      - WASTEBIN_DB_DRIVER=sqlite3           # Set -db-driver flag
      - WASTEBIN_DB_SOURCE=/data/lenpaste.db # Set -db-source flag
      - WASTEBIN_DB_CLEANUP_PERIOD=3h        # Set -db-cleanup-period flag
      - WASTEBIN_ROBOTS_DISALLOW=false       # If true set -robots-disallow flag
      - WASTEBIN_TITLE_MAX_LENGTH=100        # Set -title-max-length flag. If 0 disable title, if -1 disable length limit.
      - WASTEBIN_BODY_MAX_LENGTH=10000       # Set -body-max-length flag. If -1 disable length limit. Can't be -1.
      - WASTEBIN_MAX_PASTE_LIFETIME=never    # Set -max-paste-lifetime flag. Examples: 2d, 12h, 7m.
      - WASTEBIN_ADMIN_NAME=                 # Set -admin-name flag.
      - WASTEBIN_ADMIN_MAIL=                 # Set -admin-mail flag.
    volumes:
      # /data/lenpaste.db - SQLite DB
      # /data/about.html  - About this server
      # /data/rules.html  - This server rules
      - "${PWD}/data:/data"
      - "/etc/timezone:/etc/timezone:ro"
      - "/etc/localtime:/etc/localtime:ro"
    ports:
      - "80:80"
```

3. Execute while in the directory where `docker-compose.yml` is located:

```bash
docker-compose pull && docker-compose up -d
```

PS: If you want to install updates, run:

```bash
docker-compose pull && docker-compose stop && docker-compose up -d
```

## Build from source code

On Debian/Ubuntu:

```bash
sudo apt update
sudo apt -y install git make gcc golang
git clone https://git.lcomrade.su/root/lenpaste.git
cd ./lenpaste/
make
```

You can find the result of the build in the `./dist/` directory.

## Other documentation

For instance administrators:

- [Reverse proxy: Nginx](docs/reverse_proxy_nginx.md)
- [Database: PostgreSQL](docs/db_postgresql.md)

For developers:

- [Lenpaste API](https://paste.lcomrade.su/docs/apiv1)
- [Libraries for working with API](https://paste.lcomrade.su/docs/api_libs)

## Bugs and Suggestion

If you find a bug or have a suggestion, please open an issue or pull request


CHANGELOG.md
# Changelog

Semantic versioning is used (<https://semver.org/>).

## v1.1

- You can now specify author, author email and author URL for paste.
- Full localization into Russian.
- Added settings menu.
- Paste creation and expiration times are now displayed in the user's time zone.
- Add PostgreSQL DB support.

## v1.0

This is the first stable release of Lenpaste🎉

Compared to the previous unstable versions, everything has been drastically improved:
design, loading speed of the pages, API, work with the database.
Plus added syntax highlighting in the web interface.

## v0.2

Features:

- Paste title
- About server information
- Improved documentation
- Logging and log rotation
- Storage configuration
- Code optimization

Bug fixes:

- Added `./version.json` to Docker image
- Added paste expiration check before opening
- Fixed incorrect error of expired pastes
- API errors now return in JSON

## v0.1

Features:

- Alternative to pastebin.com
- Creating expiration pastes
- Web interface
- API


cmd/main.go
package main

import (
	"errors"
	"flag"
	"io/ioutil"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"time"

	"github.com/coolguy1771/wastebin/internal/apiv1"
	"github.com/coolguy1771/wastebin/internal/config"
	"github.com/coolguy1771/wastebin/internal/logger"
	"github.com/coolguy1771/wastebin/internal/raw"
	"github.com/coolguy1771/wastebin/internal/storage"
	"github.com/coolguy1771/wastebin/internal/web"
)

var Version = ""

func backgroundJob(cleanJobPeriod time.Duration, db storage.DB, log logger.Config) {
	for {
		// Delete expired pastes
		count, err := db.PasteDeleteExpired()
		if err != nil {
			log.Error(errors.New("Delete expired: " + err.Error()))
		}

		log.Info("Delete " + strconv.FormatInt(count, 10) + " expired pastes")

		// Wait
		time.Sleep(cleanJobPeriod)
	}
}

func readFile(path string) (string, error) {
	// Open file
	file, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer file.Close()

	// Read file
	fileByte, err := ioutil.ReadAll(file)
	if err != nil {
		return "", err
	}

	// Return result
	return string(fileByte), nil
}

func exitOnError(e error) {
	println("error:", e.Error())
	os.Exit(1)
}

func printHelp(noErrors bool) {
	println("Usage:", os.Args[0], "[-web-dir] [OPTION]...")
	println("")
	println("  -addres             ADDRES:PORT (default: :80)")
	println("  -web-dir            Dir with page templates and static content")
	println("  -db-driver          Currently supported drivers: 'sqlite3' and 'postgres' (default: sqlite3)")
	println("  -db-source          DB source")
	println("  -db-cleanup-period  Interval at which the DB is cleared of expired but not yet deleted pastes. (default: 3h)")
	println("  -robots-disallow    Prohibits search engine crawlers from indexing site using robots.txt file.")
	println("  -title-max-length   Maximum length of the paste title. If 0 disable title, if -1 disable length limit. (default: 100)")
	println("  -body-max-length    Maximum length of the paste body. If -1 disable length limit. Can't be -1. (default: 100000)")
	println("  -max-paste-lifetime Maximum lifetime of the paste. Examples: 12h, 7m, 10s. (default: never)")
	println("  -server-about       Path to the HTML file that contains the server description.")
	println("  -server-rules       Path to the HTML file that contains the server rules.")
	println("  -admin-name         Name of the administrator of this server")
	println("  -admin-mail         Email of the administrator of this server.")
	println("  -version            Display version and exit")
	println("  -help               Display this help and exit")
	println()
	println("Exit status:")
	println(" 0  if you used the -help or -version flag")
	println(" 1  if there is an error during initialization")
	println(" 2  if command line arguments are not entered correctly")

	if noErrors == false {
		os.Exit(2)
	}

	os.Exit(0)
}

func printVersion() {
	println(Version)

	os.Exit(0)
}

func printFlagNotSet(flg string) {
	println("flag is not set:", flg)

	os.Exit(2)
}

func init() {
	if Version == "" {
		Version = "unknown"
	}
}

func main() {
	// Get ./bin/ dir
	binFile, err := os.Executable()
	if err != nil {
		exitOnError(err)
	}

	binFile, err = filepath.EvalSymlinks(binFile)
	if err != nil {
		exitOnError(err)
	}

	binDir := filepath.Dir(binFile)

	// Get ./share/lenpaste/web dir
	defaultWebDir := filepath.Join(binDir, "../share/lenpaste/web")

	// Read cmd args
	flag.Usage = func() { printHelp(false) }

	flagAddress := flag.String("address", ":80", "")
	flagWebDir := flag.String("web-dir", defaultWebDir, "")
	flagDbDriver := flag.String("db-driver", "sqlite3", "")
	flagDbSource := flag.String("db-source", "", "")
	flagDbCleanupPeriod := flag.String("db-cleanup-period", "3h", "")
	flagRobotsDisallow := flag.Bool("robots-disallow", false, "")
	flagTitleMaxLen := flag.Int("title-max-length", 100, "")
	flagBodyMaxLen := flag.Int("body-max-length", 10000, "")
	flagMaxLifetime := flag.String("max-paste-lifetime", "never", "")
	flagServerAbout := flag.String("server-about", "", "")
	flagServerRules := flag.String("server-rules", "", "")
	flagAdminName := flag.String("admin-name", "", "")
	flagAdminMail := flag.String("admin-mail", "", "")
	flagVersion := flag.Bool("version", false, "")
	flagHelp := flag.Bool("help", false, "")

	flag.Parse()

	// -help flag
	if *flagHelp == true {
		printHelp(true)
	}

	// -version flag
	if *flagVersion == true {
		printVersion()
	}

	// -db-source flag
	if *flagDbSource == "" {
		printFlagNotSet("-db-source")
	}

	// -body-max-length flag
	if *flagBodyMaxLen == 0 {
		println("-body-max-length flag cannot be 0")
		os.Exit(2)
	}

	// -max-paste-lifetime
	maxLifeTime := int64(-1)

	if *flagMaxLifetime != "never" {
		maxLifeTimeTmp, err := time.ParseDuration(*flagMaxLifetime)
		if err != nil {
			exitOnError(err)
		}

		maxLifeTime = int64(maxLifeTimeTmp.Seconds())

		if maxLifeTime < 600 {
			println("-max-paste-lifetime flag cannot have a value less than 10 minutes")
			os.Exit(2)
		}
	}

	// Load server about
	serverAbout := ""
	if *flagServerAbout != "" {
		serverAbout, err = readFile(*flagServerAbout)
		if err != nil {
			exitOnError(err)
		}
	}

	// Load server rules
	serverRules := ""
	if *flagServerRules != "" {
		serverRules, err = readFile(*flagServerRules)
		if err != nil {
			exitOnError(err)
		}
	}

	// Settings
	db := storage.DB{
		DriverName:     *flagDbDriver,
		DataSourceName: *flagDbSource,
	}

	log := logger.Config{
		TimeFormat: "2006/01/02 15:04:05",
	}

	cfg := config.Config{
		DB:             db,
		Log:            log,
		Version:        Version,
		TitleMaxLen:    *flagTitleMaxLen,
		BodyMaxLen:     *flagBodyMaxLen,
		MaxLifeTime:    maxLifeTime,
		ServerAbout:    serverAbout,
		ServerRules:    serverRules,
		AdminName:      *flagAdminName,
		AdminMail:      *flagAdminMail,
		RobotsDisallow: *flagRobotsDisallow,
	}

	apiv1Data := apiv1.Load(cfg)

	rawData := raw.Load(cfg)

	// Init data base
	err = db.InitDB()
	if err != nil {
		exitOnError(err)
	}

	// Load pages
	webData, err := web.Load(cfg, *flagWebDir)
	if err != nil {
		exitOnError(err)
	}

	// Handlers
	http.HandleFunc("/robots.txt", func(rw http.ResponseWriter, req *http.Request) {
		webData.RobotsTxtHand(rw, req)
	})
	if *flagRobotsDisallow == false {
		http.HandleFunc("/sitemap.xml", func(rw http.ResponseWriter, req *http.Request) {
			webData.SitemapHand(rw, req)
		})
	}

	http.HandleFunc("/style.css", func(rw http.ResponseWriter, req *http.Request) {
		webData.StyleCSSHand(rw, req)
	})

	http.HandleFunc("/", func(rw http.ResponseWriter, req *http.Request) {
		webData.MainHand(rw, req)
	})
	http.HandleFunc("/main.js", func(rw http.ResponseWriter, req *http.Request) {
		webData.MainJSHand(rw, req)
	})
	http.HandleFunc("/paste.js", func(rw http.ResponseWriter, req *http.Request) {
		webData.PasteJSHand(rw, req)
	})

	http.HandleFunc("/settings", func(rw http.ResponseWriter, req *http.Request) {
		webData.SettingsHand(rw, req)
	})

	http.HandleFunc("/raw/", func(rw http.ResponseWriter, req *http.Request) {
		rawData.RawHand(rw, req)
	})
	http.HandleFunc("/dl/", func(rw http.ResponseWriter, req *http.Request) {
		webData.DlHand(rw, req)
	})
	http.HandleFunc("/emb/", func(rw http.ResponseWriter, req *http.Request) {
		webData.EmbeddedHand(rw, req)
	})
	http.HandleFunc("/emb_help/", func(rw http.ResponseWriter, req *http.Request) {
		webData.EmbeddedHelpHand(rw, req)
	})

	http.HandleFunc("/about", func(rw http.ResponseWriter, req *http.Request) {
		webData.AboutHand(rw, req)
	})
	http.HandleFunc("/about/license", func(rw http.ResponseWriter, req *http.Request) {
		webData.LicenseHand(rw, req)
	})
	http.HandleFunc("/about/source_code", func(rw http.ResponseWriter, req *http.Request) {
		webData.SourceCodePageHand(rw, req)
	})

	http.HandleFunc("/docs", func(rw http.ResponseWriter, req *http.Request) {
		webData.DocsHand(rw, req)
	})
	http.HandleFunc("/docs/apiv1", func(rw http.ResponseWriter, req *http.Request) {
		webData.DocsApiV1Hand(rw, req)
	})
	http.HandleFunc("/docs/api_libs", func(rw http.ResponseWriter, req *http.Request) {
		webData.DocsApiLibsHand(rw, req)
	})

	http.HandleFunc("/api/", func(rw http.ResponseWriter, req *http.Request) {
		apiv1Data.MainHand(rw, req)
	})
	http.HandleFunc("/api/v1/new", func(rw http.ResponseWriter, req *http.Request) {
		apiv1Data.NewHand(rw, req)
	})
	http.HandleFunc("/api/v1/get", func(rw http.ResponseWriter, req *http.Request) {
		apiv1Data.GetHand(rw, req)
	})
	http.HandleFunc("/api/v1/getServerInfo", func(rw http.ResponseWriter, req *http.Request) {
		apiv1Data.GetServerInfoHand(rw, req)
	})

	// Run background job
	jobDuration, err := time.ParseDuration(*flagDbCleanupPeriod)
	if err != nil {
		exitOnError(err)
	}

	go backgroundJob(jobDuration, db, log)

	// Run HTTP server
	log.Info("Run HTTP server on " + *flagAddress)
	err = http.ListenAndServe(*flagAddress, nil)
	if err != nil {
		exitOnError(err)
	}
}
